Add focus metric

DominionSoftware opened the pull request:
I added the focus metric code. It looks at the taken image and adds a red frame and text around the image if the focus metric is less than 950.0. I wasn't able to get my tablet to be out of focus very easily, so I added Unit test code (commented out) for testing.

I added strings to the resources and used google translate for swahili.

Elforama commented on app/src/main/java/org/greenstand/android/TreeTracker/activities/CameraActivity.kt:
```diff
@@ -198,10 +206,109 @@ class CameraActivity : AppCompatActivity(), Camera.PictureCallback, View.OnClick
+        val rows = input.lastIndex + 1;
+        val cols = input.get(0).lastIndex + 1;
+
+        var V = Array(rows) { Array(cols) { 0 } }
```
Can we please use lower case for variable names?

Elforama commented on app/src/main/java/org/greenstand/android/TreeTracker/activities/CameraActivity.kt:
```diff
@@ -198,10 +206,109 @@ class CameraActivity : AppCompatActivity(), Camera.PictureCallback, View.OnClick
+        for (row in 0 until rows) {
+            for (col in 0 until cols) {
+                val HRC = H[row][col];
+                val VRC = V[row][col];
```
Same as above

Elforama commented on app/src/main/java/org/greenstand/android/TreeTracker/fragments/NewTreeFragment.kt:
```diff
@@ -225,6 +233,17 @@ class NewTreeFragment : androidx.fragment.app.Fragment(), OnClickListener, Activ
                     }
 
                     setPic()
+                    if (MainActivity.mImageQuality < FOCUS_THRESHOLD) {
```
Instead of adding a new property into the MainActivty, can we instead pass back the image quality value using onActivityResult?

Elforama commented on app/src/main/java/org/greenstand/android/TreeTracker/fragments/NewTreeFragment.kt:
```diff
@@ -85,6 +88,11 @@ class NewTreeFragment : androidx.fragment.app.Fragment(), OnClickListener, Activ
 
+        frameLayout = v.fragment_new_tree_frame_layout
+
+        focusWarning = v.fragment_new_tree_focus_warning_text
```
These local references to the view are not needed. We can directly use fragment_new_tree_frame_layout and fragment_new_tree_focus_warning_text without using `v`

DominionSoftware replied:
Ok this is done.

DominionSoftware commented:
mImageView already existed so I took that as a model.

How does one do this ?

From: Jonathan Muller <notifications@github.com>
Sent: Tuesday, June 04, 2019 8:29 PM
To: Greenstand/treetracker-android <treetracker-android@noreply.github.com>
Cc: Richard Frank <rfrank@dominionsw.com>; Author <author@noreply.github.com>
Subject: Re: [Greenstand/treetracker-android] Add focus metric (#322)


@Elforama commented on this pull request.

________________________________

In app/src/main/java/org/greenstand/android/TreeTracker/fragments/NewTreeFragment.kt<https://github.com/Greenstand/treetracker-android/pull/322#discussion_r290545042>:

> @@ -85,6 +88,11 @@ class NewTreeFragment : androidx.fragment.app.Fragment(), OnClickListener, Activ



         mImageView = v.fragmentNewTreeImage



+        frameLayout = v.fragment_new_tree_frame_layout

+

+        focusWarning = v.fragment_new_tree_focus_warning_text

These local references to the view are not needed. We can directly use fragment_new_tree_frame_layout and fragment_new_tree_focus_warning_text without using v

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Greenstand/treetracker-android/pull/322?email_source=notifications&email_token=AETPBYQSE2VYZKRIRP2DR3LPY4CE5A5CNFSM4HTHS5IKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2TA4IA#pullrequestreview-245763616>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AETPBYQRN3WE4RDS42D7GDTPY4CE5ANCNFSM4HTHS5IA>.

Elforama commented:
Any place where you use focusWarning, replace it with fragment_new_tree_focus_warning_text. And any place you use frameLayout, replace it with fragment_new_tree_frame_layout. Those long underscore names are part of a plugin called Kotlin Synthetic Views. It allows us to directly reference views from a xml layout. You can read more about it here if you're interested: https://kotlinlang.org/docs/tutorials/android-plugin.html

Feel free to update the imageView to do the same.

Elforama commented on app/src/main/java/org/greenstand/android/TreeTracker/activities/CameraActivity.kt:
```diff
@@ -328,6 +434,98 @@ class CameraActivity : AppCompatActivity(), Camera.PictureCallback, View.OnClick
 
 
+/** UNIT TEST CODE FOR FOCUS METRIC
+    private fun printImage(image: Array<Array<Int>> ) {
```
What do you think of moving this testing code into its own file?

DominionSoftware replied:
I move everything to ImageUtils

DominionSoftware commented:
That's fine. Where should the file live? Are there other unit test like files?




Rick Frank
Dominion Software, Inc.
825 Beacon Street
Newton, MA 02459
Medical, Scientific, and Industrial Software

On Jun 4, 2019, at 9:33 PM, Jonathan Muller <notifications@github.com<mailto:notifications@github.com>> wrote:


@Elforama commented on this pull request.

________________________________

In app/src/main/java/org/greenstand/android/TreeTracker/activities/CameraActivity.kt<https://github.com/Greenstand/treetracker-android/pull/322#discussion_r290554752>:

> @@ -328,6 +434,98 @@ class CameraActivity : AppCompatActivity(), Camera.PictureCallback, View.OnClick
     }


+/** UNIT TEST CODE FOR FOCUS METRIC
+    private fun printImage(image: Array<Array<Int>> ) {


What do you think of moving this testing code into its own file?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Greenstand/treetracker-android/pull/322?email_source=notifications&email_token=AETPBYUBAMWEAXEFPCIOX2TPY4JXLA5CNFSM4HTHS5IKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2TDZQQ#pullrequestreview-245775554>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AETPBYXUMGTEUFWFHA3QZ3LPY4JXLANCNFSM4HTHS5IA>.

DominionSoftware commented:
OK I moved all the image processing code to ImageUtils.

Elforama commented:
@DominionSoftware Looks like there are some minor conflicts with the strings file.

DominionSoftware commented:
Ok string conflict resolved

Elforama commented:
@DominionSoftware I've tried running this on my Pixel 3, but once I take a picture the app freezes for a while before crashing. Any idea what's going on?

DominionSoftware commented:
No idea without more information. I guess I need to get the phone from Ezra. ( he's sending me one ) I only have a samsung tablet to test with.


Rick Frank
Dominion Software, Inc.
825 Beacon Street
Newton, MA 02459
Medical, Scientific, and Industrial Software

On Jun 9, 2019, at 12:00 PM, Jonathan Muller <notifications@github.com<mailto:notifications@github.com>> wrote:


@DominionSoftware<https://github.com/DominionSoftware> I've tried running this on my Pixel 3, but once I take a picture the app freezes for a while before crashing. Any idea what's going on?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Greenstand/treetracker-android/pull/322?email_source=notifications&email_token=AETPBYSRBNOPBGKVNABCPZLPZUSJPA5CNFSM4HTHS5IKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXINCZY#issuecomment-500224359>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AETPBYWJDJ5FZYUMALQDVNLPZUSJPANCNFSM4HTHS5IA>.

Elforama commented:
I just got it working with my emulator, however it's always showing the image as blurry even when I've ensured the image was sharp. Are you seeing this behavior at all?

Elforama commented on build.gradle:
```diff
@@ -7,7 +7,7 @@ buildscript {
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.3.1'
+        classpath 'com.android.tools.build:gradle:3.4.1'
```
This has caused some build issues, can we change it back?

DominionSoftware replied:
I don't have any reason to not change it back

DominionSoftware replied:
The way I tested it was to blur images using the gaussian filter I put into ImageUtiles and then measure the metric compared to non blurred images, the result being empirically tested to be metric of 950.0

DominionSoftware replied:
![image](https://user-images.githubusercontent.com/19329250/59161801-0d65cd80-8ab5-11e9-99e2-c1d6e6851b8a.png)
Current UI without trash can

Elforama commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.1-all.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-5.1.1-all.zip
```
Same as above

Elforama commented:
Here is the user flow for blur detection: https://docs.google.com/drawings/d/1CkkffY48rDrCJ3hLVtdVa4diHrFUj5D3oAzlLWuU_Ec/edit

There doesn't yet seem to be a way to retake the photo.

@peacengn How should we allow the user to retake the photo? By tapping on the photo?

DominionSoftware commented:
> I just got it working with my emulator, however it's always showing the image as blurry even when I've ensured the image was sharp. Are you seeing this behavior at all?

No.

DominionSoftware commented:
I limited the image size for the focus test to 200 pixels wide. I changed the color to be Yellow, as a caution, rather than red. Because if you take a picture of an in focus blank wall, it will say it's out of focus (because it's low contrast). So it should be a warning / reminder and not a fail. I changed the string to be:

"We've detected that your photo may be blurry. Please push the back button retake the photo, if necessary.

I also made the focus threshold slightly lower (permissive) based on some tests. I did some tests with my tablet and Pixel 3A XL cameras and it seemed to be working as expected. Some tests with less capable phones are needed.

DominionSoftware commented:
I framed the text "Caution Yellow" instead of the image as I could not get the image to frame tightly.

The blur detection code adds no noticeable performance hit to my phone or tablet. However, they are fairly new an have sufficient RAM.


Blurred Image test (used a piece of cellophane to blur the image); non blurred image follows after.


![Screenshot_20190611-170500](https://user-images.githubusercontent.com/19329250/59311385-43eb4600-8c77-11e9-9739-c1bb9afa58b1.png)



![Screenshot_20190611-170359](https://user-images.githubusercontent.com/19329250/59311478-a2b0bf80-8c77-11e9-9def-52151f1040c5.png)

DominionSoftware commented:
Closing pull request per conversations with Ezra and Sebastian. Code can stay in branch if future use becomes desired.
